arch/arm/core/userspace.S

```diff
@@ -0,0 +1,154 @@
+/*
+ * Userspace and service handler hooks
+ *
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include <offsets_short.h>
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <kernel_structs.h>
+#include <arch/cpu.h>
+#include <syscall.h>
+
+_ASM_FILE_PROLOGUE
+
+GTEXT(_arm_userspace_enter)
+GTEXT(_arm_do_syscall)
+GDATA(_kernel)
+
+/* Imports */
+GTEXT(_k_syscall_table)
+
+/**
+ *
+ * User space entry function
+ *
+ * This function is the entry point to user mode from privileged execution.
+ * The conversion is one way, and threads which transition to user mode do
+ * not transition back later, unless they are doing system calls.
+ *
+ */
+SECTION_FUNC(TEXT,_arm_userspace_enter)
+    /* preserve arguments */
+    mov lr,r0
+    push {r1,r2,r3,lr}
+
+    /* disable mpu stack guard to keep us from triggering guard region */
+    bl arm_core_mpu_disable
+    pop {r1,r2,r3,lr}
+
+    /*
+     * calculate end of stack and set SP to it
+     */
+    ldr r0, [sp]
+    ldr ip, [sp,#4]
+    add r0,ip
+
+    msr PSP, r0
+
+    /* push args to stack, pad end with 2 words of 0x0 */
+    sub sp,8
+    push {r1,r2,r3,lr}
+    ldr r3,=0
+    str r3,[sp, #20]
+    str r3,[sp, #24]
+
+    /* clear the stack area to clean out privileged data */
+#ifdef CONFIG_INIT_STACKS
+    ldr r1,=0xaaaaaaaa
+#else
+    ldr r1,=0
+#endif
+    mov r2,ip
+    sub r2,#24		/* don't erase the pushed regs */
+    bl memset
```

Comment:
r0 will be stack_obj + stack_info.size here, so memset() will set the memory out of the range of stack_obj.

Reply:
has this been fixed?

Reply:
yeah now we switch over to the priv stack and then memset out the complete user mode stack area.  No need to dance around any stored info here.